tracer/src/Datadog.Trace/Ci/GitCommandHelper.cs

```diff
@@ -188,6 +189,222 @@ static List<FileCoverageInfo> ParseGitDiff(string diffOutput)
         }
     }
 
+    public static BaseBranchInfo? DetectBaseBranch(
+        string workingDirectory,
+        string? targetBranch = null,
+        string remoteName = "origin",
+        string branchFilterPattern = @"^(main|master|preprod|prod|release/.*|hotfix/.*)$")
+    {
+        if (string.IsNullOrEmpty(workingDirectory))
+        {
+            Log.Warning("GitCommandHelper: Cannot detect base branch because working directory is null or empty");
+            return null;
+        }
+
+        try
+        {
+            // 1. Get the target branch if not specified
+            if (string.IsNullOrEmpty(targetBranch))
+            {
+                var branchOutput = RunGitCommand(
+                    workingDirectory,
+                    "rev-parse --abbrev-ref HEAD",
+                    MetricTags.CIVisibilityCommands.GetHead);
+
+                if (branchOutput is not { ExitCode: 0 } || string.IsNullOrWhiteSpace(branchOutput?.Output))
+                {
+                    Log.Warning("GitCommandHelper: Failed to get current branch");
+                    return null;
+                }
+
+                targetBranch = branchOutput!.Output.Trim();
+            }
+
+            // Verify branch exists
+            var verifyBranchOutput = RunGitCommand(
+                workingDirectory,
+                $"rev-parse --verify --quiet {targetBranch}",
+                MetricTags.CIVisibilityCommands.VerifyBranchExists);
+
+            if (verifyBranchOutput is not { ExitCode: 0 })
+            {
+                Log.Warning("GitCommandHelper: Branch '{Branch}' does not exist", targetBranch);
+                return null;
+            }
+
+            // 2. Check if the target is already a main-like branch
+            string shortTargetName = targetBranch!;
+            if (shortTargetName.StartsWith($"{remoteName}/"))
+            {
+                shortTargetName = shortTargetName.Substring(remoteName.Length + 1);
+            }
+
+            if (Regex.IsMatch(shortTargetName, branchFilterPattern))
```

Comment:
Similar Regex over `branchFilterPattern` is used below, you could refactor and create a single `Regex` instance for both.

Reply:
Changed!